README.md
![DocParser Logo](https://github.com/stautonico/PyDocParser/blob/master/img/logo.png?raw=true)

## Unofficial Python Client for the Docparser API

### Table of Contents

* [Installation](#Installation)
* [Usage](#Usage)
* [Contributing](#Contributing)
* [License](#License)
* [Changelog](#Changelog)
* [To-Do](#To-Do)
* [Documentation](https://docparser.tautonico.tech/docs)
* [Additional Information](#Additional-Information)

<br>

### Installation

**Important Note:** PyDocParser does not support Python 2.X. **The minimum compatible version is Python 3.8.**

#### Install with pip

`pip install pydocparser`

#### Install from Source

1. Download the release of your choice from [here](https://github.com/stautonico/PyDocParser/releases)
2. Unzip the file (`unzip PyDocParser-<version>.zip` or use your favorite archive manager)
3. Change directory to the unzipped folder (`cd PyDocParser-<version>`)
4. Run `python setup.py install`

#### Install for Development

1. Clone the repository (`git clone https://github.com/stautonico/pydocparser`)
2. (Optional, but recommended) Create a virtual environment (`python -m venv venv`)
3. Activate the virtual environment (`source venv/bin/activate`)
4. Install the dependencies (`pip install -r requirements.txt`)

<br>

### Usage

This is a basic example of how to use PyDocParser. For more examples, check out
the [documentation](https://docparser.tautonico.tech/docs).

The first thing you need to do is create an instance of the `Docparser` class from the `pydocparser` module, and
call the `login()` method, passing your API key as an argument:

```python
import pydocparser

parser = pydocparser.Docparser()

login_result = parser.login("your_api_key")  # returns True if successful, False otherwise
```

Next, you can upload a document to a parser. This example uploads a local file using the `upload_local_file_by_path()`
method, providing the path to the file and the label of the parser as arguments:

```python
try:
    document_id = parser.upload_local_file_by_path("path/to/file.pdf", "parser_label")
except Exception as e:
    print("Something went wrong while uploading the file")
    print(e)
    exit(1)

print(document_id)  # The ID of the document in the parser if uploading was successful
```

Finally, you can retrieve the parsed data using the `get_one_result()` and providing the parser label and the document
ID as arguments:

```python
try:
    result = parser.get_one_result("parser_label", document_id)
except Exception as e:
    print("Something went wrong while retrieving the parsed data")
    print(e)
    exit(1)

print(result)  # The parsed data if retrieval was successful (an array of dictionaries)
```

To view a full list of the available methods, check out the [documentation](https://docparser.tautonico.tech/docs).

### Contributing

Thank you for considering contributing to **PyDocParser**. We appreciate your interest and support in making this
project
better.

There aren't many guidelines for contributing to this project, but there are a few simple rules:

- Be respectful towards other contributors and maintainers.
- Make sure your code is clean and well-documented.
- Write clear commit messages and include relevant information.
- Be civil and open to receiving feedback.

#### How To Contribute

1. Fork the repository: Start by forking this repository to your GitHub account. This will create a copy of the project
   under your account, allowing you to make changes without affecting the original codebase.

2. Clone the repository: Next, clone the repository to your local machine using the following command:

`git clone https://github.com/<your_username>/PyDocParser`

3. Create a new branch: Before making any changes, create a new branch to work on. Use a descriptive name that reflects
   the
   nature of your contribution. For example:

`git checkout -b new-feature-testing`

`git checkout -b bug-fix-123`

`git checkout -b documentation-update`

Also, if you're working on an issue, you can use the issue number as the branch name. For example:

`git checkout -b issue-123-bug`

`git checkout -b issue-456-docs`

`git checkout -b issue-789-feature`

4. Make your changes: Now, you can start making your desired changes to the project. Whether it's fixing a bug, adding a
   new feature, or improving the documentation, we appreciate all kinds of contributions.

5. Commit your changes: Once you're satisfied with your changes, commit them with a clear and concise commit message
   that
   describes the purpose of your modifications:

`git commit -m "Add my new feature"`

`git commit -m "Fix issue #123"`

`git commit -m "Update documentation (fixes #456)"`

6. Push your changes: Push your changes to your forked repository on GitHub.

`git push origin my-new-feature`

7. Open a pull request: Finally, open a pull request (PR) from your forked repository to the original repository.
   Provide a detailed description of the changes you've made, and I will review your contribution as soon as
   possible.

Thank you again for your interest and support in contributing to our project.

### License

This project is licensed under the MIT License, which is a permissive open-source license. It grants you the freedom to
use, modify, and distribute the software for any purpose. For more information, please refer to
the [LICENSE](LICENSE.md) file.

### Changelog

For a list of changes and new features, please refer to the [CHANGELOG](CHANGELOG.md) file.

### To-Do

This is just a small list of things that may be added in the future that I didn't want to create full issues for:

- [ ] Automate the generation of the documentation
- [ ] Automate the building and publishing of the package (both on PyPI and GitHub)

### Documentation

For more information, please refer to the [documentation](https://docparser.tautonico.tech/docs).

### Additional Information

This project is not affiliated with Docparser in any way. For more information about Docparser, please
  visit [their website](https://docparser.com/)

<br>

This project requires a minimum Python version of 3.8. It is designed to take advantage of the features and improvements introduced in Python 3.8 and later versions.

As of today (June 13, 2023), Python 3.7 is approaching its end-of-life (EOL) in only 14 days, on June 27, 2023. Python 2.x reached its EOL on January 1, 2020. It is strongly recommended to use a newer version of Python to ensure compatibility with the latest libraries and security patches.

Upgrading to a newer version of Python will provide you with access to the latest language features, performance enhancements, and bug fixes. It also ensures that you can take full advantage of the improvements in this project.

If you are currently using Python 2.x or Python 3.7, we encourage you to upgrade to Python 3.8 or a later version. The Python community provides comprehensive documentation and migration guides to assist you in the upgrade process.


<br>

This library has been recently updated. If you are using and older version, and need help migrating to the latest version,
please refer to the [migration guide](2.0.0->2023.06.a_migration_guide.md).

CHANGELOG.md
## Changelog

### Release 2023.06.a (June 06, 2023)
The Entire codebase was re-written because
1. It hasn't been touched in years
2. I've gotten better at writing code
3. To make all names + order of operations fit the official API
4. To support all features of the official API (some were missing before)


### Release V2.0 (6/3/2021)

### Full rewrite!
The Entire codebase was re-written because
1. It seemed to behave strangely sometimes
2. To make all names + order of operations fit the official API

### Release V1.1.2.1 (1/14/20)

### Added:
* `uploadbase64()`

### Release V1.1 (7/12/19)

#### Added:

* `find_parser_id()`
* `upload_id()`
* `fetch_id()`
* `fetch_all_id()`

#### Changed:

* Renamed `parsers()` to `get_parsers()`

pydocparser/__init__.py
__author__ = "Steve Tautonico"
__contact__ = "stautonico@gmail.com"
__date__ = "2023-06-12T00:00:00"
__version__ = "2023.06.a"

from datetime import datetime
import os

from typing import Optional, Union, List, Dict, Literal

import requests


class Docparser:
    """
    A class wrapper for the docparser API.
    This class covers all the API endpoints available from https://docparser.com/api
    """

    def __init__(self):
        self._BASE_URL = "https://api.docparser.com/v1"
        self._api_key = None

    def login(self, api_key: str) -> bool:
        """
        Login to the docparser API.
        Args:
            api_key (str): The API key to use for authentication.

        Returns:
            bool: True if login was successful, False otherwise.
        """
        self._api_key = api_key

        return self.ping()

    def _make_authenticated_request(self, method: str, url: str, data: dict = None,
                                    files: dict = None) -> requests.Response:
        """
        Make an authenticated request to the docparser API.
        Args:
            method (str): The HTTP method to use.
            url (str): The URL to make the request to.
            data (dict): The data to send with the request.
            files (dict): The files to send with the request.

        Returns:
            requests.Response: The response object.
        """
        return requests.request(method, url, data=data, files=files, auth=(self._api_key, ""))

    def ping(self) -> bool:
        """
        Ping the docparser API to test the connection.
        (wrapper for https://api.docparser.com/v1/ping)
        Returns:
            bool: True if the connection was successful, False otherwise.
        """
        url = self._BASE_URL + "/ping"
        response = self._make_authenticated_request("GET", url)
        if response.status_code == 200:
            # Just confirm that our response looks like "{"msg": "pong"}"
            return response.json().get("msg") == "pong"
        else:
            return False

    def _get_parser_id_by_label(self, parser_label: str) -> Optional[str]:
        """
        Get the parser ID by the parser label.
        Args:
            parser_label (str): The parser label to search for.

        Returns:
            Optional[str]: The parser ID if found, None otherwise.
        """
        try:
            parsers = self.list_parsers()
        except Exception as e:
            raise e

        for parser in parsers:
            if parser.get("label") == parser_label:
                return parser.get("id")

        return None

    def _check_parser_id_exists(self, parser_label: str) -> Optional[str]:
        """
        Check if the parser ID exists.
        Args:
            parser_label (str): The parser label to search for.

        Returns:
            Optional[str]: The parser ID if found, None otherwise.
        """
        try:
            parser_id = self._get_parser_id_by_label(parser_label)
        except Exception as e:
            raise e

        if parser_id is None:
            raise Exception(f"Parser label \"{parser_label}\" not found.")

        return parser_id

    # Parser Routes
    def list_parsers(self) -> Union[List[dict], str]:
        """
        Get a list of all document parsers linked to the current account.
        (wrapper for https://api.docparser.com/v1/parsers)

        Returns:
            Optional[Union[List[dict], str]]: A list of all document parsers linked to the current account, or an exception if the request failed including the error message.
        """
        result = self._make_authenticated_request("GET", self._BASE_URL + "/parsers")

        if result.status_code == 200:
            if type(result.json()) == list:
                return result.json()

        raise Exception(result.text)

    def list_parser_model_layouts(self, parser_label: str) -> Union[List[dict], str]:
        """
        Get a list of all the Model Layouts for a specific parser by label.
        (wrapper for https://api.docparser.com/v1/parser/models/<PARSER_ID>)
        Args:
            parser_label (str): The parser label to search for.

        Returns:
            Optional[Union[List[dict], str]]: A list of all the Model Layouts for a specific parser by label, or an exception if the request failed including the error message.
        """
        parser_id = self._check_parser_id_exists(parser_label)

        result = self._make_authenticated_request("GET", self._BASE_URL + "/parser/models/" + parser_id)

        if result.status_code == 200:
            if type(result.json()) == list:
                return result.json()

        raise Exception(result.text)

    # Document Routes
    def upload_local_file_by_path(self, path: str, parser_label: str, remote_id: Optional[str] = None) -> str:
        """
        Upload a local file (by file path) to a specific parser (by label)
        (wrapper for https://api.docparser.com/v1/document/upload/<PARSER_ID>)
        Args:
            path (str): The file path to upload.
            parser_label (str): The parser label to upload to.
            remote_id (str): The ID of the remote file being uploaded. (optional)

        Returns:
            Optional[str]: The document ID if successful, raises an exception otherwise.
        """
        parser_id = self._check_parser_id_exists(parser_label)

        path = os.path.expanduser(path)

        # Check if the file exists
        if not os.path.exists(path):
            raise Exception(f"File path {path} does not exist.")

        files = {"file": open(path, "rb")}
        result = self._make_authenticated_request("POST", self._BASE_URL + "/document/upload/" + parser_id,
                                                  data={"remote_id": remote_id}, files=files)

        if result.status_code == 200:
            return result.json().get("id")

        raise Exception(result.text)

    def upload_file_by_base64(self, base64_file_content: bytes, filename: str, parser_label: str,
                              remote_id: Optional[str] = None) -> str:
        """
        Upload a file by base64 encoded content to a specific parser (by label)
        (wrapper for https://api.docparser.com/v1/document/upload/<PARSER_ID>)
        Args:
            base64_file_content (bytes): The base64 encoded file content.
            filename (str): The filename of the file being uploaded.
            parser_label (str): The parser label to upload to.
            remote_id (str): The ID of the remote file being uploaded. (optional)

        Returns:
            Optional[str]: The document ID if successful, raises an exception otherwise.
        """
        parser_id = self._check_parser_id_exists(parser_label)

        payload = {
            "file_content": base64_file_content,
            "file_name": filename,
            "remote_id": remote_id
        }

        result = self._make_authenticated_request("POST", self._BASE_URL + "/document/upload/" + parser_id,
                                                  data=payload)

        if result.status_code == 200:
            return result.json().get("id")

        raise Exception(result.text)

    def upload_file_by_url(self, url: str, parser_label: str, remote_id: Optional[str] = None) -> str:
        """
        Upload a file (by public URL) to a specific parser (by label)
        (wrapper for https://api.docparser.com/v1/document/fetch/<PARSER_ID>)
        Args:
            url (str): The public URL of the file to upload.
            parser_label (str): The parser label to upload to.
            remote_id (str): The ID of the remote file being uploaded. (optional)

        Returns:
            Optional[str]: The document ID if successful, raises an exception otherwise.
        """
        parser_id = self._check_parser_id_exists(parser_label)

        payload = {
            "url": url,
            "remote_id": remote_id
        }

        result = self._make_authenticated_request("POST", self._BASE_URL + "/document/fetch/" + parser_id,
                                                  data=payload)

        if result.status_code == 200:
            return result.json().get("id")

        raise Exception(result.text)

    def get_one_result(self, parser_label: str, document_id: str, include_children: Optional[bool] = False) -> List[
        Dict[str, Union[str, int]]]:
        """
        Get the parsed result of a single document by ID.
        (wrapper for https://api.docparser.com/v1/results/<PARSER_ID>/<DOCUMENT_ID>)
        Note that the "format" option is currently not supported.
        Args:
            parser_label (str): The parser label to search for.
            document_id (str): The document ID to search for.
            include_children (bool): Whether to include the children of the result. (optional)
                                     Examples include documents created while splitting documents, etc.
        Returns:
            Optional[List[Dict[str, Union[str, int]]]]: The parsed result of a single document by ID, raises an exception otherwise.
        """
        parser_id = self._check_parser_id_exists(parser_label)

        payload = {
            "include_children": include_children,
        }

        result = self._make_authenticated_request("POST", self._BASE_URL + "/results/" + parser_id + "/" + document_id,
                                                  data=payload)

        if result.status_code == 200:
            return result.json()

        raise Exception(result.text)

    def get_multiple_results(self, parser_label: str,
                             _list: Literal["last_uploaded", "uploaded_after", "processed_after"] = "last_uploaded",
                             limit: int = 100, date: datetime = None, remote_id: Optional[str] = None,
                             include_processing_queue: bool = False, sort_by: Literal[
                "parsed_at", "processed_at", "uploaded_at", "first_processed_at", "imported_at", "integrated_at", "dispatched_webook_at", "preprocessed_at"] = "uploaded_at",
                             sort_order: Literal["asc", "desc"] = "desc") -> List[Dict[str, Union[str, int]]]:
        """
        Get multiple results from a parser by label.
        (wrapper for https://api.docparser.com/v1/results/<PARSER_ID>
        Note: The "format" option is currently not supported.
        Args:
            parser_label (str): The parser label to search for.
            _list (Literal["last_uploaded", "uploaded_after", "processed_after"]): The type of documents to list. (optional)
            limit (int): The maximum number of results to return (if list is == "last_uploaded".) (optional)
            date (datetime): Used for "uploaded_after" and "processed_after" (optional, mandatory if _list is "uploaded_after" or "processed_after")
            remote_id (str): The remote ID to search for. (optional)
            include_processing_queue (bool): Whether to include files that are still processing (optional)
            sort_by (Literal["parsed_at", "processed_at", "uploaded_at", "first_processed_at", "imported_at", "integrated_at", "dispatched_webook_at", "preprocessed_at"]): The field to sort by. (optional)
            sort_order (Literal["asc", "desc"]): The sort order. (optional)
        """
        parser_id = self._check_parser_id_exists(parser_label)

        if _list == "uploaded_after" or _list == "processed_after":
            if date is None:
                raise Exception("date must be provided if _list is 'uploaded_after' or 'processed_after'")

            date = date.isoformat()

        payload = {
            "list": _list,
            "limit": limit,
            "date": date,
            "remote_id": remote_id,
            "include_processing_queue": include_processing_queue,
            "sort_by": sort_by,
            "sort_order": sort_order.upper()
        }

        result = self._make_authenticated_request("POST", self._BASE_URL + "/results/" + parser_id, data=payload)

        if result.status_code == 200:
            return result.json()

        raise Exception(result.text)
